### PR TITLE
raw formatter and improved duration calculation

### PR DESCRIPTION
**Pull request by AaronJhaj**
- Added raw formatter and unit tests
- created a new time parsing function to get more accurate duration
- moved some code from the formatters into runsFormatter code
- IVT test for raw formatter
- refactored time-related code in formatters

### CODE REVIEW

File: pkg/formatters/rawFormatter.go [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+func NewRawFormatter() RunsFormatter {
+	return new(RawFormatter)
+}
+
+func (*RawFormatter) GetName() string {
+	return RAW_FORMATTER_NAME
+}
+
+func (*RawFormatter) IsNeedingDetails() bool {
+	return false
```
Comment (techcobweb) [inline pkg/formatters/rawFormatter.go:31]: It should need all the details surely ?

**Comment by galasa-team** [comment]
Build successful

File: pkg/runs/runsGet.go [hunk 2]
```diff
@@ -73,6 +73,9 @@ func createFormatters() map[string]formatters.RunsFormatter {
 	detailedFormatter := formatters.NewDetailsFormatter()
 	validFormatters[detailedFormatter.GetName()] = detailedFormatter
 
+	rawFormatter := formatters.NewRawFormatter()
```
Comment (techcobweb) [inline pkg/runs/runsGet.go:76]: so simple to plug in now we did the leg work.

File: pkg/formatters/runsFormatter.go [hunk 3]
```diff
@@ -32,3 +40,35 @@ func calculateMaxLengthOfEachColumn(table [][]string) []int {
...
+	return parsedTime
+}
+
+func calculateDurationMilliseconds(start time.Time, end time.Time) string {
+	duration := strconv.FormatInt(end.Sub(start).Milliseconds(), 10)
+
+	return duration
+}
+
+// func calculateDurationMilliseconds(startTimeString string, endTimeString string) string {
```
[the remark this answers is not in the document]
Comment (AaronJhaj) [reply pkg/formatters/runsFormatter.go:63]: ah yes forgot to remove this. will remove

**Review by techcobweb — APPROVED** [review]
A few minor things, but looks great to me if you can sort those out.

File: pkg/formatters/rawFormatter.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+		var duration string = ""
+
+		var startTimeStringForDuration time.Time
+		var endTimeStringForDuration time.Time
+
+		startTimeStringRaw := run.TestStructure.GetStartTime()
+		endTimeStringRaw := run.TestStructure.GetEndTime()
+
+		if len(startTimeStringRaw) > 0 {
```
Comment (techcobweb) [inline pkg/formatters/rawFormatter.go:49]: Is this code duplicated with the other formatters which work out time deltas ?

**Comment by galasa-team** [comment]
Build successful

File: pkg/formatters/rawFormatter.go (hunk 4 above)
Comment (AaronJhaj) [reply pkg/formatters/rawFormatter.go:49]: Yes it is similar to code in the details formatter. I will refactor

File: pkg/runs/runsGet.go [hunk 5]
```diff
@@ -47,7 +47,7 @@ func GetRuns(
 			if err == nil {
 
 				// Some formatters need extra fields filled-in so they can be displayed.
-				if chosenFormatter.IsNeedingDetails() {
+				if chosenFormatter.IsNeedingMethodDetails() {
 					runJson, err = GetRunDetailsFromRasSearchRuns(runJson, apiServerUrl)
```
Comment (techcobweb) [inline pkg/runs/runsGet.go:51]: Why isn't the method detail available already from the original REST call ? Seems that it possibly should be returned in all cases from the query.

File: test-galasactl-ecosystem.sh [hunk 6]
```diff
@@ -332,6 +333,44 @@ function runs_get_check_details_format_output {
...
+    cat $output_file | grep "${GALASA_TEST_NAME_LONG}" -q
+    rc=$?
+    # We expect a return code of '0' because the test name should be output.
+    if [[ "${rc}" != "0" ]]; then 
+        error "Did not find ${GALASA_TEST_NAME_LONG} in raw output"
+        exit 1
+    fi  
+
+    # Check that we got 10 pipes
+    pipe_count=$(grep -o "|" $output_file | wc -l | xargs)
```
Comment (techcobweb) [inline test-galasactl-ecosystem.sh:364]: that's not getting a pipe count. It's counting the lines which have pipes in. wc -l counts lines.

**Review by techcobweb — APPROVED** [review]
Approved, though I do suggest a few minor things.

**Comment by galasa-team** [comment]
Build failed, see http://localhost:8001/api/v1/namespaces/tekton-pipelines/services/tekton-dashboard:http/proxy/#/namespaces/galasa-build/pipelineruns/repo-cli-pr-94-5vzg9 for details. If you are unable to do so, please contact a member of the Galasa team.

File: pkg/formatters/detailsFormatter.go [hunk 7]
```diff
@@ -120,25 +105,13 @@ func initialiseMethodTable() [][]string {
...
-			startTimeStringReadable = formatTimeReadable(startTimeStringRaw)
-			startTimeForDuration = formatTimeForDurationCalculation(startTimeStringRaw)
-			if len(endTimeStringRaw) > 0 {
-				endTimeStringReadable = formatTimeReadable(endTimeStringRaw)
-				endTimeForDuration = formatTimeForDurationCalculation(endTimeStringRaw)
-				duration = calculateDurationMilliseconds(startTimeForDuration, endTimeForDuration)
-			}
-		}
+		startTimeStringReadable := getReadableTime(startTimeStringRaw)
+		endTimeStringReadable := getReadableTime(endTimeStringRaw)
```
Comment (techcobweb) [inline pkg/formatters/detailsFormatter.go:112]: nice

File: pkg/formatters/runsFormatter.go [hunk 8]
```diff
@@ -59,3 +59,27 @@ func calculateDurationMilliseconds(start time.Time, end time.Time) string {
 
 	return duration
 }
+
+func getDuration(startTimeStringRaw string, endTimeStringRaw string) string {
```
Comment (techcobweb) [inline pkg/formatters/runsFormatter.go:63]: nice

**Comment by galasa-team** [comment]
Build successful

**Pull request merged**